src/GUI/UI/Houses/Shops.cpp

```diff
@@ -775,33 +867,47 @@ void GUIWindow_Shop::houseSpecificDialogue() {
-            GenerateStandartShopItems();
-            GenerateSpecialShopItems();
+            generateShopItems(false);
+            generateShopItems(true);
             GameTime nextGenTime = pParty->GetPlayingTime() + GameTime::FromDays(buildingTable[wData.val - 1].generation_interval_days);
             pParty->PartyTimes.shopNextRefreshTime[houseId()] = nextGenTime;
         }
 
-        std::array<ItemGen, 12> &itemArray = (option == DIALOGUE_SHOP_BUY_STANDARD) ? pParty->standartItemsInShops[houseId()] : pParty->specialItemsInShops[houseId()];
-        if (ItemAmountForShop(buildingTable[wData.val - 1].uType)) {
-            for (int i = 0; i < ItemAmountForShop(buildingTable[wData.val - 1].uType); ++i) {
-                if (itemArray[i].uItemID != ITEM_NULL) {
-                    shop_ui_items_in_store[i] = assets->getImage_ColorKey(itemArray[i].GetIconName());
-                }
+        const std::array<ItemGen, 12> &itemArray = (option == DIALOGUE_SHOP_BUY_STANDARD) ? pParty->standartItemsInShops[houseId()] : pParty->specialItemsInShops[houseId()];
+        for (int i = 0; i < itemAmountInShop[buildingTable[wData.val - 1].uType]; ++i) {
+            if (itemArray[i].uItemID != ITEM_NULL) {
+                shop_ui_items_in_store[i] = assets->getImage_ColorKey(itemArray[i].GetIconName());
             }
         }
-        if (_buildingType == BuildingType_WeaponShop) {
-            if (ItemAmountForShop(buildingTable[wData.val - 1].uType)) {
-                for (int i = 0; i < ItemAmountForShop(buildingTable[wData.val - 1].uType); ++i) {
-                    if (itemArray[i].uItemID != ITEM_NULL) {
-                        // Note that we're using grng here for a reason - we want recorded mouse clicks to work.
-                        weaponYPos[i] = grng->random(300 - shop_ui_items_in_store[i]->height());
-                    }
+        if (buildingTable[wData.val - 1].uType == BuildingType_WeaponShop) {
+            for (int i = 0; i < itemAmountInShop[buildingTable[wData.val - 1].uType]; ++i) {
+                if (itemArray[i].uItemID != ITEM_NULL) {
+                    // Note that we're using grng here for a reason - we want recorded mouse clicks to work.
+                    weaponYPos[i] = grng->random(300 - shop_ui_items_in_store[i]->height());
                 }
             }
         }
+    } else if (IsSkillLearningDialogue(option)) {
+        learnSelectedSkill(GetLearningDialogueSkill(option));
+    }
+}
+
+std::vector<DIALOGUE_TYPE> GUIWindow_Shop::listDialogueOptions(DIALOGUE_TYPE option) {
+    switch (dialog_menu_id) {
+      case DIALOGUE_MAIN:
+        return {DIALOGUE_SHOP_BUY_STANDARD, DIALOGUE_SHOP_BUY_SPECIAL, DIALOGUE_SHOP_DISPLAY_EQUIPMENT, DIALOGUE_LEARN_SKILLS};
+      case DIALOGUE_SHOP_DISPLAY_EQUIPMENT:
+        if (buildingTable[wData.val - 1].uType == BuildingType_AlchemistShop) {
+            return {DIALOGUE_SHOP_SELL, DIALOGUE_SHOP_IDENTIFY};
```

Comment:
That's some questionable encapsulation.

On the one hand, we have `GUIWindow_AlchemyShop`. On the other hand, we peek into the actual shop type in the base class.

Reply:
Did the 3rd option.